File: core/src/main/java/computer/living/gamepadyn/InputEventListener.java
package computer.living.gamepadyn;

@FunctionalInterface
public interface InputEventListener<
    T extends InputData,
    TD extends Enum<TD> & ActionEnumDigital,
    TA extends Enum<TA> & ActionEnumAnalog1,
    TAA extends Enum<TAA> & ActionEnumAnalog2>
{
    void accept(Event.EventData<T, TD, TA, TAA> it);
}

File: core/src/main/kotlin/computer/living/gamepadyn/InputEventListener.kt
package computer.living.gamepadyn

import computer.living.gamepadyn.Event.EventData

/**
 * Functional interface for Gamepadyn event listeners. See [Event]
 * @see InputEventListener.onStateChange
 */
@FunctionalInterface
fun interface InputEventListener<T : InputData, TD, TA, TAA>
        where TD : Enum<TD>,
              TD : ActionEnumDigital,
              TA : Enum<TA>,
              TA : ActionEnumAnalog1,
              TAA : Enum<TAA>,
              TAA : ActionEnumAnalog2
{
    /**
     * Called when an action's state changes.
     * @param ev The relevant event data.
     */
    fun onStateChange(ev: EventData<T, TD, TA, TAA>)
}